helix-term/src/commands/engine/scheme.rs

```diff
@@ -0,0 +1,2003 @@
+                                inner.borrow().as_ref(),
+                            ) {
+                                return Some(value.clone());
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        if let SteelVal::Boxed(boxed_map) = REVERSE_BUFFER_MAP.with(|x| x.clone()) {
+            if let SteelVal::HashMapV(map) = boxed_map.borrow().clone() {
+                if let Some(label) = map.get(&SteelVal::IntV(document_id_to_usize(doc_id) as isize))
+                {
+                    if let SteelVal::Boxed(boxed_map) =
+                        BUFFER_OR_EXTENSION_KEYBINDING_MAP.with(|x| x.clone())
+                    {
+                        if let SteelVal::HashMapV(map) = boxed_map.borrow().clone() {
+                            if let Some(value) = map.get(label) {
+                                if let SteelVal::Custom(inner) = value {
+                                    if let Some(_) =
+                                        steel::rvals::as_underlying_type::<EmbeddedKeyMap>(
+                                            inner.borrow().as_ref(),
+                                        )
+                                    {
+                                        return Some(value.clone());
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Refer to the global keybinding map for the rest
+        Some(GLOBAL_KEYBINDING_MAP.with(|x| x.clone()))
+    }
+}
+
+pub fn initialize_engine() {
+    ENGINE.with(|x| x.borrow().globals().first().copied());
+}
+
+pub fn compositor_present_error(cx: &mut compositor::Context, e: SteelErr) {
+    cx.editor.set_error(format!("{}", e));
+
+    let backtrace = ENGINE.with(|x| x.borrow_mut().raise_error_to_string(e));
+
+    let callback = async move {
+        let call: job::Callback = Callback::EditorCompositor(Box::new(
+            move |editor: &mut Editor, compositor: &mut Compositor| {
+                if let Some(backtrace) = backtrace {
+                    let contents = ui::Markdown::new(
+                        format!("```\n{}\n```", backtrace),
+                        editor.syn_loader.clone(),
+                    );
+                    ui::Text::new(format!("```\n{}\n```", backtrace));
+                    let popup = Popup::new("engine", contents).position(Some(
+                        helix_core::Position::new(editor.cursor().0.unwrap_or_default().row, 2),
+                    ));
+                    compositor.replace_or_push("engine", popup);
+                }
+            },
+        ));
+        Ok(call)
+    };
+    cx.jobs.callback(callback);
+}
+
+pub fn present_error_inside_engine_context(cx: &mut Context, engine: &mut Engine, e: SteelErr) {
```

Comment:
same review comments as below, probablly shoud share this code too instead of duplicating it (the other method should jsut call this one)

Reply:
done